python/ray/autoscaler/_private/autoscaler.py

```diff
@@ -478,50 +504,65 @@ def _get_nodes_allowed_to_terminate(
                 # No resources of the node were needed for request_resources().
                 # max_node_resources[i] is an empty dict for legacy yamls
                 # before the node is connected.
-                nodes_allowed_to_terminate[node_id] = True
+                pass
             else:
-                nodes_allowed_to_terminate[node_id] = False
-        return nodes_allowed_to_terminate
+                nodes_not_allowed_to_terminate.add(node_id)
+        return frozenset(nodes_not_allowed_to_terminate)
 
     def _keep_worker_of_node_type(self, node_id: NodeID,
                                   node_type_counts: Dict[NodeType, int]
-                                  ) -> Tuple[bool, Dict[NodeType, int]]:
+                                  ) -> Tuple[KeepOrTerminate, Optional[str]]:
         """Determines if a worker should be kept based on the min_workers
-        constraint of the worker's node_type.
+        and max_workers constraint of the worker's node_type.
 
-        Returns True exactly when both of the following hold:
+        Returns KeepOrTerminate.keep when both of the following hold:
         (a) The worker's node_type is present among the keys of the current
             config's available_node_types dict.
         (b) Deleting the node would violate the min_workers constraint for that
             worker's node_type.
 
-        Also updates and returns the dictionary of node type counts.
+        Returns KeepOrTerminate.terminate when either the node type is not
+        among available node type, or both the following hold:
+        (a) The worker's node_type is present among the keys of the current
```

Comment:
(a) should read "is not present"
(a) is redundant with wording immediately above

Reply:
thanks